File: 01.two-sum/TwoSum.go
package main

import "fmt"

func twoSums(arr []int, sum int) (int, int) {
	numMap := make(map[int]int, len(arr))
	for i, val := range arr {
		if j, ok := numMap[sum-val]; ok {
			return j, i
		}
		numMap[val] = i
	}
	return -1, -1
}

func solveProblem(arr []int, sum int) {
	i, j := twoSums(arr, sum)
	fmt.Printf("array: %v, expect sum: %d, indexes: [%d, %d] --> number(%d, %d)",
		arr, sum, i, j, arr[i], arr[j])
}

func main() {
	solveProblem([]int{2, 7, 11, 15}, 9)
}

File: 01.two-sum/TwoSum.py

def two_sum(arr: list[int], sum: int) -> tuple[int, int]:
    num_index = {}
    for i, value in enumerate(arr):
        if sum - value in num_index:
            return num_index[sum - value], i
        num_index[value] = i
    return -1, -1


def solve_problem(arr: list[int], sum: int) -> None:
    i, j = two_sum(arr, sum)
    print(f"array: {arr}, sum: {sum}, indexes: [{i}, {j}], values: ({arr[i]}, {arr[j]})")


if __name__ == '__main__':
    solve_problem([2, 7, 11, 15], 9)
    solve_problem([2, 7, 11, 15], 26)
